packages/router/utils.ts

```diff
@@ -135,29 +135,58 @@ export interface ShouldRevalidateFunction {
 }
 
 /**
- * A route object represents a logical route, with (optionally) its child
- * routes organized in a tree-like structure.
+ * Base RouteObject with common props shared by all types of routes
  */
-export interface AgnosticRouteObject {
+type AgnosticBaseRouteObject = {
   caseSensitive?: boolean;
-  children?: AgnosticRouteObject[];
-  index?: boolean;
   path?: string;
   id?: string;
   loader?: LoaderFunction;
   action?: ActionFunction;
   hasErrorBoundary?: boolean;
   shouldRevalidate?: ShouldRevalidateFunction;
   handle?: any;
-}
+};
 
 /**
- * A data route object, which is just a RouteObject with a required unique ID
+ * Index routes must not have children
+ */
+export type AgnosticIndexRouteObject = AgnosticBaseRouteObject & {
+  children?: undefined;
+  index: true;
+};
+
+/**
+ * Non-index routes may have children, but cannot have index
+ */
+export type AgnosticNonIndexRouteObject = AgnosticBaseRouteObject & {
+  children?: AgnosticRouteObject[];
+  index?: false;
+};
```

Comment:
These are the new foundation of the types.  They get extended in react-router with `element`/`errorElement` and then again in `RouteProps` to change children from `RouteObject` to `ReactNode`